lib/galaxy/datatypes/microarrays.py

```diff
@@ -0,0 +1,152 @@
+
+
+def count_special_lines(word, filename, invert=False):
+    """
+        searching for special 'words' using the grep tool
+        grep is used to speed up the searching and counting
+        The number of hits is returned.
+    """
+    try:
+        cmd = ["grep", "-c"]
+        if invert:
+            cmd.append('-v')
+        cmd.extend([word, filename])
+        out = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+        return int(out.communicate()[0].split()[0])
+    except Exception:
+        pass
+    return 0
+
+
+def count_lines(filename, non_empty=False):
+    """
+        counting the number of lines from the 'filename' file
+    """
+    try:
+        out = subprocess.Popen(['wc', '-l', filename], stdout=subprocess.PIPE)
+        return int(out.communicate()[0].split()[0])
+    except Exception:
+        pass
+    return 0
+
+
+class GenericMicroarrayFile(data.Text):
+    """
+    Abstract class for most of the microarray files.
+    """
+    MetadataElement(name="number_of_comments", default=1, desc="Number of comments", readonly=True, visible=True,
+                    optional=True, no_value=0)
+    MetadataElement(name="number_of_blocks", default=1, desc="Number of blocks", readonly=True, visible=True,
+                    optional=True, no_value=0)
+    MetadataElement(name="number_of_records", default=1, desc="Number of records", readonly=True, visible=True,
+                    optional=True, no_value=0)
+
+    def set_peek(self, dataset, is_multi_byte=False):
+        if not dataset.dataset.purged:
+            if dataset.metadata.number_of_blocks == 1:
+                dataset.blurb = "1 block %s recodrs" % dataset.metadata.number_of_records
+                dataset.peek = get_n_first_lines(dataset.file_name, n=3)
+            else:
+                dataset.blurb = "%s blocks %s recodrs" % (dataset.metadata.number_of_blocks, dataset.metadata.number_of_records)
+                dataset.peek = get_n_first_lines(dataset.file_name, n=3)
+        else:
+            dataset.peek = 'file does not exist'
+            dataset.blurb = 'file purged from disk'
+
+    def get_mime(self):
+        return 'text/plain'
+
+
+class Gal(GenericMicroarrayFile):
+    """ Gal File format described at:
+            http://mdc.custhelp.com/app/answers/detail/a_id/18883/#gal
+    """
+
+    edam_format = "format_3829"
+    edam_data = "data_3110"
+    file_ext = "gal"
+
+    def sniff(self, filename):
+        """
+        Try to guess if the file is a Gal file.
+        >>> from galaxy.datatypes.sniff import get_test_fname
+        >>> fname = get_test_fname('test.gal')
+        >>> Gal().sniff(fname)
+        True
+        >>> fname = get_test_fname('test.gpr')
+        >>> Gal().sniff(fname)
+        False
+        """
+        found_genepix = count_special_lines("GenePix", filename) >= 1
+        found_atf = count_special_lines("ATF", filename) >= 1
+        found_blockcount = count_special_lines("BlockCount", filename) >= 1
```

Comment:
I feat that forking three grep subprocesses when sniffing is going to be a big performance hit.

Can you please have a look at how sniffing is done using the `sniff_prefix` method in `lib/galaxy/datatypes/molecules.py` ?

Reply:
I did look at molecules file very deeply. The get_peek doesn't support well, mixed encoding files. The case of Gpr, it contains a special character that io.string fails to deal with. I committed the changes you suggested, but the test will probably fail.

Reply:
@nsoranzo Gpr is an ISO-8859 and Gal is ASCII.

Reply:
ISO-8859 is a legacy encoding, I've locally converted the GPR test file from ISO-8859-1 to UTF-8 using `recode` (see e.g. https://linux-tips.com/t/converting-files-between-various-character-sets-and-usages/161 ) and this fixes the tests.

Do you think that's something acceptable for your use case? If yes, I can upload here the converted test file.

Reply:
Better leave it as it is. The tools i built will not convert, they need just this type to be detected by galaxy while filtering inputs. But great idea thank you

Reply:
Then the tests won't pass at this point, and we can't merge it. It is an issue we'll think about and work on, but I'd suggest converting the test files to a standard encoding for the time being. I also agree with @nsoranzo, we'll not want to run external commands during sniffing.

Reply:
@nsoranzo has already converted the files (Thanks for that). I can get rid of the external then. thank you both @mvdbeek  and @nsoranzo.

Reply:
Are you sure, test.gpr still contains "Rgn R�" on line 29, which looks like a non-UT8 character.

Reply:
I've converted it on my laptop, I'll push it later.

Reply:
> I fear that forking three grep subprocesses when sniffing is going to be a big performance hit.
>
> Can you please have a look at how sniffing is done using the sniff_prefix method in `lib/galaxy/datatypes/molecules.py` ?

This is still an issue, can you have another look now that the test file is fixed?

Reply:
I will update it tomorrow morning